Translate: cloneElement (#680)
reactjs/ko.react.dev: src/content/reference/react/cloneElement.md

console.log(clonedElement); // <Row title="Cabbage">Goodbye</Row>
```

[아래에서 더 많은 예제를 볼 수 있습니다.](#usage)

#### 매개변수 {/*parameters*/}

> * `element`: `element` 인자는 유효한 React 엘리먼트여야 합니다. 예를 들어, `<Something />`과 같은 JSX 노드, [`createElement`](/reference/react/createElement)로 호출한 결과물 또는 다른 `cloneElement`로 호출한 결과물이 될 수 있습니다.

Reviewer: 제안) 풀어 번역하면 뜻이 명징할 것 같습니다.

```suggestion
* `element`: `element` 인자는 유효한 React 엘리먼트여야 합니다. 예를 들어, `<Something />`과 같은 JSX 노드, [`createElement`](/reference/react/createElement)를 호출해 얻은 결과물 또는 다른 `cloneElement`로 호출해 얻은 결과물이 될 수 있습니다.
```

* `props`: `props` 인자는 객체 또는 `null`이어야 합니다. `null`을 전달하면 복제된 엘리먼트는 원본 `element.props`를 모두 유지합니다. 그렇지 않으면 `props` 객체의 각 prop에 대해 반환된 엘리먼트는 `element.props`의 값보다 `props`의 값을 "우선"합니다. 나머지 `props`는 원본 `element.props`에서 채워집니다. `props.key` 또는 `props.ref`를 전달하면 원본의 것을 대체합니다.

* **(선택사항)** `...children`: 0개 이상의 자식 노드가 필요합니다. React 엘리먼트, 문자열, 숫자, [portals](/reference/react-dom/createPortal), 빈 노드 (`null`, `undefined`, `true`, `false`) 및 React 노드 배열을 포함한 모든 React 노드가 해당할 수 있습니다. `...children` 인자를 전달하지 않으면 원본 `element.props.children`이 유지됩니다.

> 

Reviewer: 불필요한 개행이 있습니다

#### 반환값 {/*returns*/}

`cloneElement`는 다음과 같은 프로퍼티를 가진 React 엘리먼트 객체를 반환합니다.

* `type`: `element.type`과 동일합니다.
* `props`: `element.props`와 전달한 `props`를 얕게 병합한 결과입니다.
* `ref`: `props.ref`에 의해 재정의되지 않은 경우 원본 `element.ref`입니다.
* `key`: `props.key`에 의해 재정의되지 않은 경우 원본 `element.key`입니다.

> 일반적으로 컴포넌트에서 엘리먼트를 반환하거나 다른 엘리먼트의 자식으로 만듭니다. 엘리먼트의 프로퍼티를 읽을 수 있지만, 생성된 후에는 모든 엘리먼트를 불투명하게 취급하고 렌더링하는 것이 좋습니다.

Reviewer: 제안) "불투명" 이런 직역이 이해를 살짝 어렵게 하는 감이 있어서 의역하면 어떨까 싶습니다.

```suggestion
일반적으로 컴포넌트에서 엘리먼트를 반환하거나 다른 엘리먼트의 자식으로 만듭니다. 엘리먼트의 프로퍼티를 읽을 수 있지만, 생성된 후에는 모든 엘리먼트의 프로퍼티를 읽을 수 없는 것처럼 취급하고 렌더링하는 것이 좋습니다.
```

#### 주의 {/*caveats*/}

* 엘리먼트를 복제해도 **원본 엘리먼트는 수정되지 않습니다.**

> * **자식이 모두 정적으로 알려진 경우에만** `cloneElement(element, null, child1, child2, child3)`와 같이 **자식을 여러 개의 인자로 전달해야 합니다.** 자식이 동적으로 생성되었다면 `cloneElement(element, null, listItems)`와 같이 전체 배열을 세 번째 인자로 전달해야 합니다. 이렇게 하면 React가 모든 동적 리스트에 대해 [key가 누락되었다는 경고](/learn/rendering-lists#keeping-list-items-in-order-with-key)를 보여줍니다. 정적 리스트의 경우는 순서가 변경되지 않으므로 이 작업은 필요하지 않습니다.

Reviewer: 직역하지 않아도 뜻이 통할 것 같습니다.

```suggestion
* **자식이 모두 정적인 경우에만** `cloneElement(element, null, child1, child2, child3)`와 같이 **자식을 여러 개의 인자로 전달해야 합니다.** 자식이 동적으로 생성되었다면 `cloneElement(element, null, listItems)`와 같이 전체 배열을 세 번째 인자로 전달해야 합니다. 이렇게 하면 React가 모든 동적 리스트에 대해 [key가 누락되었다는 경고](/learn/rendering-lists#keeping-list-items-in-order-with-key)를 보여줍니다. 정적 리스트의 경우는 순서가 변경되지 않으므로 이 작업은 필요하지 않습니다.
```

* `cloneElement`는 데이터 흐름을 추적하기 어렵기 때문에 다음 [대안](#alternatives)을 사용해 보세요.

---

> ## 사용 {/*usage*/}

Reviewer: 제안) "사용법" 정도가 단락 제목에 맞는 번역이 아닐까 생각합니다.

```suggestion
## 사용법 {/*usage*/}
```

</Sandpack>

[context를 통해 데이터를 전달하는 것에 대하여 자세히 알아보세요.](/reference/react/useContext#passing-data-deeply-into-the-tree)

---

### Custom Hook으로 로직 추출하기 {/*extracting-logic-into-a-custom-hook*/}

> 다른 접근 방식으로는 자체 hook을 통해 "비ㅠ시각적인" 로직을 추출하는 것을 시도해 볼 수 있습니다. 그리고 hook에 의해서 반환된 정보를 사용하여 렌더링할 내용을 정합니다. 예를 들어 다음과 같이 `useList` 같은 custom hook을 작성할 수 있습니다.

Reviewer 2: ```suggestion
다른 접근 방식으로는 자체 hook을 통해 "비시각적인" 로직을 추출하는 것을 시도해 볼 수 있습니다. 그리고 hook에 의해서 반환된 정보를 사용하여 렌더링할 내용을 정합니다. 예를 들어 다음과 같이 `useList` 같은 custom hook을 작성할 수 있습니다.
```

Reviewer 3: 요건 제가 반영하고 머지하겠습니다 🙇‍♂️

Author: 앗..! 놓쳤네요ㅠ 감사합니다 ㅎㅎ

</Sandpack>

[context를 통해 데이터를 전달하는 것에 대하여 자세히 알아보세요.](/reference/react/useContext#passing-data-deeply-into-the-tree)

---

### Custom Hook으로 로직 추출하기 {/*extracting-logic-into-a-custom-hook*/}

> 다른 접근 방식으로는 자체 hook을 통해 "비 시각적인" 로직을 추출하는 것을 시도해 볼 수 있습니다. 그리고 hook에 의해서 반환된 정보를 사용하여 렌더링할 내용을 정합니다. 예를 들어 다음과 같이 `useList` 같은 custom hook을 작성할 수 있습니다.

Reviewer: ```suggestion
다른 접근 방식으로는 자체 hook을 통해 "비시각적인" 로직을 추출하는 것을 시도해 볼 수 있습니다. 그리고 hook에 의해서 반환된 정보를 사용하여 렌더링할 내용을 정합니다. 예를 들어 다음과 같이 `useList` 같은 custom hook을 작성할 수 있습니다.
```
